// @amazon-web-services-cloudformation/cloudformation-cli-typescript-lib/dist/interface.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnmodeledRequest = exports.BaseResourceHandlerRequest = exports.HandlerRequest = exports.RequestData = exports.TestEvent = exports.BaseModel = exports.BaseDto = exports.HandlerErrorCode = exports.OperationStatus = exports.MetricTypes = exports.StandardUnit = exports.Action = exports.Integer = void 0;
require("reflect-metadata");
const class_transformer_1 = require("class-transformer");
exports.Integer = new Proxy(BigInt, {
    apply(target, _thisArg, argArray) {
        target.prototype.toJSON = function () {
            return Number(this.valueOf());
        };
        const isSafeInteger = (value) => {
            if (value &&
                (value < BigInt(Number.MIN_SAFE_INTEGER) ||
                    value > BigInt(Number.MAX_SAFE_INTEGER))) {
                return false;
            }
            return true;
        };
        target.isSafeInteger = isSafeInteger;
        const value = target(...argArray);
        if (value && !isSafeInteger(value)) {
            throw new RangeError(`Value is not a safe integer: ${value.toString()}`);
        }
        return value;
    },
});
var Action;
(function (Action) {
    Action["Create"] = "CREATE";
    Action["Read"] = "READ";
    Action["Update"] = "UPDATE";
    Action["Delete"] = "DELETE";
    Action["List"] = "LIST";
})(Action = exports.Action || (exports.Action = {}));
var StandardUnit;
(function (StandardUnit) {
    StandardUnit["Count"] = "Count";
    StandardUnit["Milliseconds"] = "Milliseconds";
})(StandardUnit = exports.StandardUnit || (exports.StandardUnit = {}));
var MetricTypes;
(function (MetricTypes) {
    MetricTypes["HandlerException"] = "HandlerException";
    MetricTypes["HandlerInvocationCount"] = "HandlerInvocationCount";
    MetricTypes["HandlerInvocationDuration"] = "HandlerInvocationDuration";
})(MetricTypes = exports.MetricTypes || (exports.MetricTypes = {}));
var OperationStatus;
(function (OperationStatus) {
    OperationStatus["Pending"] = "PENDING";
    OperationStatus["InProgress"] = "IN_PROGRESS";
    OperationStatus["Success"] = "SUCCESS";
    OperationStatus["Failed"] = "FAILED";
})(OperationStatus = exports.OperationStatus || (exports.OperationStatus = {}));
var HandlerErrorCode;
(function (HandlerErrorCode) {
    HandlerErrorCode["NotUpdatable"] = "NotUpdatable";
    HandlerErrorCode["InvalidRequest"] = "InvalidRequest";
    HandlerErrorCode["AccessDenied"] = "AccessDenied";
    HandlerErrorCode["InvalidCredentials"] = "InvalidCredentials";
    HandlerErrorCode["AlreadyExists"] = "AlreadyExists";
    HandlerErrorCode["NotFound"] = "NotFound";
    HandlerErrorCode["ResourceConflict"] = "ResourceConflict";
    HandlerErrorCode["Throttling"] = "Throttling";
    HandlerErrorCode["ServiceLimitExceeded"] = "ServiceLimitExceeded";
    HandlerErrorCode["NotStabilized"] = "NotStabilized";
    HandlerErrorCode["GeneralServiceException"] = "GeneralServiceException";
    HandlerErrorCode["ServiceInternalError"] = "ServiceInternalError";
    HandlerErrorCode["NetworkFailure"] = "NetworkFailure";
    HandlerErrorCode["InternalFailure"] = "InternalFailure";
    HandlerErrorCode["InvalidTypeConfiguration"] = "InvalidTypeConfiguration";
})(HandlerErrorCode = exports.HandlerErrorCode || (exports.HandlerErrorCode = {}));
class BaseDto {
    constructor(partial) {
        if (partial) {
            Object.assign(this, partial);
        }
    }
    serialize(removeNull = true) {
        const data = JSON.parse(JSON.stringify(class_transformer_1.classToPlain(this)));
        if (removeNull) {
            for (const key in data) {
                const value = data[key];
                if (value == null) {
                    delete data[key];
                }
            }
        }
        return data;
    }
    static deserialize(jsonData, options = {}) {
        if (jsonData == null) {
            return null;
        }
        return class_transformer_1.plainToClass(this, jsonData, {
            enableImplicitConversion: false,
            excludeExtraneousValues: true,
            ...options,
        });
    }
    toJSON(key) {
        return this.serialize();
    }
}
BaseDto.serializer = {
    classToPlain: class_transformer_1.classToPlain,
    plainToClass: class_transformer_1.plainToClass,
};
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], BaseDto.prototype, "serialize", null);
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Object)
], BaseDto.prototype, "toJSON", null);
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Object)
], BaseDto, "serializer", void 0);
exports.BaseDto = BaseDto;
class BaseModel extends BaseDto {
    getTypeName() {
        return Object.getPrototypeOf(this).constructor.TYPE_NAME;
    }
}
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", String)
], BaseModel.prototype, "getTypeName", null);
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", String)
], BaseModel, "TYPE_NAME", void 0);
exports.BaseModel = BaseModel;
class TestEvent extends BaseDto {
}
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], TestEvent.prototype, "credentials", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], TestEvent.prototype, "action", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], TestEvent.prototype, "request", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], TestEvent.prototype, "callbackContext", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], TestEvent.prototype, "region", void 0);
exports.TestEvent = TestEvent;
class RequestData extends BaseDto {
}
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "resourceProperties", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], RequestData.prototype, "providerLogGroupName", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], RequestData.prototype, "logicalResourceId", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "systemTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "stackTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "callerCredentials", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "providerCredentials", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "previousResourceProperties", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "previousStackTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], RequestData.prototype, "typeConfiguration", void 0);
exports.RequestData = RequestData;
class HandlerRequest extends BaseDto {
}
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "action", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "awsAccountId", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "bearerToken", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "region", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", RequestData)
], HandlerRequest.prototype, "requestData", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "responseEndpoint", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "stackId", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "resourceType", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "resourceTypeVersion", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], HandlerRequest.prototype, "callbackContext", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], HandlerRequest.prototype, "nextToken", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], HandlerRequest.prototype, "requestContext", void 0);
exports.HandlerRequest = HandlerRequest;
class BaseResourceHandlerRequest extends BaseDto {
}
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "clientRequestToken", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], BaseResourceHandlerRequest.prototype, "desiredResourceState", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], BaseResourceHandlerRequest.prototype, "previousResourceState", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], BaseResourceHandlerRequest.prototype, "desiredResourceTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], BaseResourceHandlerRequest.prototype, "previousResourceTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", Object)
], BaseResourceHandlerRequest.prototype, "systemTags", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "awsAccountId", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "awsPartition", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "logicalResourceIdentifier", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "nextToken", void 0);
__decorate([
    class_transformer_1.Expose(),
    __metadata("design:type", String)
], BaseResourceHandlerRequest.prototype, "region", void 0);
exports.BaseResourceHandlerRequest = BaseResourceHandlerRequest;
class UnmodeledRequest extends BaseResourceHandlerRequest {
    static fromUnmodeled(obj) {
        return UnmodeledRequest.deserialize(obj);
    }
    static getPartition(region) {
        if (!region) {
            return null;
        }
        if (region.startsWith('cn')) {
            return 'aws-cn';
        }
        if (region.startsWith('us-gov')) {
            return 'aws-gov';
        }
        return 'aws';
    }
    toModeled(modelTypeReference) {
        const request = BaseResourceHandlerRequest.deserialize({
            clientRequestToken: this.clientRequestToken,
            desiredResourceTags: this.desiredResourceTags,
            previousResourceTags: this.previousResourceTags,
            systemTags: this.systemTags,
            awsAccountId: this.awsAccountId,
            logicalResourceIdentifier: this.logicalResourceIdentifier,
            nextToken: this.nextToken,
            region: this.region,
            awsPartition: UnmodeledRequest.getPartition(this.region),
        });
        request.desiredResourceState = modelTypeReference.deserialize(this.desiredResourceState || {});
        request.previousResourceState = modelTypeReference.deserialize(this.previousResourceState || {});
        return request;
    }
}
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", BaseResourceHandlerRequest)
], UnmodeledRequest.prototype, "toModeled", null);
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", UnmodeledRequest)
], UnmodeledRequest, "fromUnmodeled", null);
__decorate([
    class_transformer_1.Exclude(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], UnmodeledRequest, "getPartition", null);
exports.UnmodeledRequest = UnmodeledRequest;
